### PR TITLE
vtk: Generate egg-info

### PR DESCRIPTION
**Pull request by bgamari**
###### Description of changes

VTK's cmake build system fails to generate an `.egg-info` file, which
various libraries in the Python package system (e.g. `pkg_resources`)
would use to identify the package. Work around this by generating an
`egg-info` file ourselves.

This allows us to also drop a workaround in `mayavi`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by wegank** [comment]
@ofborg build python3Packages.mayavi

**Review by tfmoraes — APPROVED** [review]

**Review by wegank — APPROVED** [review]

File: pkgs/development/libraries/vtk/vtk.egg-info [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Metadata-Version: 2.1
+Version: @VTK_VER@
+Summary: VTK is an open-source toolkit for 3D computer graphics, image processing, and visualization
+Platform: UNKNOWN
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/vtk/vtk.egg-info:4]: Is this field important? We could fill it with nix.

File: pkgs/development/libraries/vtk/generic.nix [hunk 2]
```diff
@@ -89,6 +89,13 @@ in stdenv.mkDerivation rec {
     sed -i 's/fprintf(output, shift)/fprintf(output, "%s", shift)/g' ./ThirdParty/libxml2/vtklibxml2/xpath.c
   '';
 
+  postInstall = optionalString enablePython ''
+    substitute \
+      ${./vtk.egg-info} \
+      $out/lib/python${pythonInterpreter.pythonVersion}/site-packages/vtk-${version}.egg-info \
+      --subst-var-by VTK_VER "${version}"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/vtk/generic.nix:96]:
```suggestion
    substitute ${./vtk.egg-info} \
      $out/${pythonInterpreter.sitePackages}/vtk-${version}.egg-info \
      --subst-var-by VTK_VER "${version}"
```


also pythonInterpreter should be renamed to python

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/vtk/vtk.egg-info (hunk 1 above)
Comment (bgamari) [reply pkgs/development/libraries/vtk/vtk.egg-info:4]: I doubt it; regardless `UNKNOWN` is the value I have seen in other `egg-info` files in the wild so I would feel better not deviating.

File: pkgs/development/libraries/vtk/generic.nix [hunk 3]
```diff
@@ -89,6 +89,12 @@ in stdenv.mkDerivation rec {
     sed -i 's/fprintf(output, shift)/fprintf(output, "%s", shift)/g' ./ThirdParty/libxml2/vtklibxml2/xpath.c
   '';
 
+  postInstall = optionalString enablePython ''
+    substitute ${./vtk.egg-info} \
+      $out/${python.sitePackages}/vtk-${version}.egg-info \
```
Comment (wegank) [inline pkgs/development/libraries/vtk/generic.nix:94]: `pythonInterpreter` should be replaced by `python` in the whole file. Also in `python-packages.nix` please replace `pythonInterpreter = python;` with `inherit (self) python;`.

**Comment by bcdarwin** [comment]
@ofborg build python310Packages.mayavi

**Review by bcdarwin — APPROVED** [review]

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
Needs a rebase.

**Review by bcdarwin — APPROVED** [review]

**Comment by wegank** [comment]
@ofborg build python3Packages.mayavi

**Review by thiagokokada — APPROVED** [review]

File: pkgs/development/libraries/vtk/generic.nix [hunk 4]
```diff
@@ -89,6 +89,13 @@ in stdenv.mkDerivation rec {
     sed -i 's/fprintf(output, shift)/fprintf(output, "%s", shift)/g' ./ThirdParty/libxml2/vtklibxml2/xpath.c
   '';
 
+  postInstall = optionalString enablePython ''
+    substitute \
+      ${./vtk.egg-info} \
+      $out/lib/python${python.pythonVersion}/site-packages/vtk-${version}.egg-info \
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/vtk/generic.nix:95]: This should use python.sitePackages

Comment (bgamari) [reply pkgs/development/libraries/vtk/generic.nix:95]: Addressed in #209971.

**Pull request merged**